### PR TITLE
Add dynamic shipping tax calculation

### PR DESCRIPTION
**Pull request by BorisovskiP**
Please make sure these boxes are checked before submitting your PR - thank you!

- [ ] Pull request is based against develop branch
- [ ] README.md reflects changes (if applicable)
- [ ] New files contain a license header

### Issue

This PR fixes issue https://github.com/firegento/firegento-magesetup2/issues/22 .

### Proposed changes

...

### CODE REVIEW

File: i18n/de_DE.csv [hunk 1]
```diff
@@ -55,3 +55,5 @@ SWIFT,SWIFT
 "CMS Page for Shipping Info","CMS-Seite für Versandkosten"
 "Show ""incl. Shipping Cost"" instead of ""excl. Shipping Cost""","Zeige ""inkl. Versandkosten"" anstatt ""exkl. Versandkosten"""
 "Display Delivery Time on Product Listing","Lieferzeit auf Produktübersichtsseiten anzeigen"
+"No dynamic shipping tax calculation","Keine dynamische Versandsteuerberechnung"
```
Comment (sprankhub) [inline i18n/de_DE.csv:58]: Please re-generate the English language file and then edit the German file. Hint:

    n98-magerun2.phar i18n:collect-phrases -o app/code/FireGento/MageSetup/i18n/en_US.csv app/code/FireGento/MageSetup/

File: composer.json [hunk 2]
```diff
@@ -2,10 +2,16 @@
   "name": "firegento/magesetup2",
   "description": "MageSetup provides the necessary configuration (system config, tax, agreements, etc. for a national market.",
   "require": {
-    "php": "~7.2.0||~7.3.0",
-    "magento/module-store": "*",
-    "magento/module-backend": "*",
-    "magento/framework": "*"
+    "php": "~7.1.0|~7.2.0|~7.3.0",
```
Comment (sprankhub) [inline composer.json:5]: Please revert this to `~7.2.0||~7.3.0`.

File: Plugin/Email/Model/Source/Variables.php [hunk 3]
```diff
@@ -8,7 +8,9 @@
 namespace FireGento\MageSetup\Plugin\Email\Model\Source;
 
 /**
- * Plugin to add additional config variables to be inserted into emails.
+ * Class Variables
```
Comment (sprankhub) [inline Plugin/Email/Model/Source/Variables.php:11]: Please revert this change.

File: Model/Config.php [hunk 4]
```diff
@@ -13,6 +13,9 @@
  */
 class Config implements ConfigInterface
 {
+    public const DYNAMIC_TYPE_DEFAULT = 0;
```
Comment (sprankhub) [inline Model/Config.php:16]: I think these should be more specific (saying that this is related to the shipping tax). Additionally, what do you think of putting them to the source model directly?

File: Model/System/Config/Source/Tax/Dynamic.php [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+use FireGento\MageSetup\Model\Config as FireGentoConfig;
+use Magento\Framework\Data\OptionSourceInterface;
+
+class Dynamic implements OptionSourceInterface
+{
+    /**
+     * @var array
+     */
+    protected $options;
```
Comment (sprankhub) [inline Model/System/Config/Source/Tax/Dynamic.php:17]: Switch to `private`.

File: Plugin/Tax/Config/ShippingTaxPlugin.php [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+use Magento\Framework\App\Config\ScopeConfigInterface;
+use Magento\Quote\Model\Quote\Item;
+use Magento\Store\Model\ScopeInterface;
+use Magento\Store\Model\Store;
+use Magento\Tax\Model\Calculation\Proxy;
+use Magento\Tax\Model\Config;
+
+class ShippingTaxPlugin
+{
+    public const CONFIG_PATH_DYNAMIC_SHIPPING_TAX_CLASS = 'tax/classes/dynamic_shipping_tax_class';
```
Comment (sprankhub) [inline Plugin/Tax/Config/ShippingTaxPlugin.php:20]: Please add a new method in the `Config` model for retrieving this configuration and use it.

File: Plugin/Tax/Config/ShippingTaxPlugin.php [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+     * @param Session              $customerSession
+     * @param GroupRepository      $groupRepository
+     * @param Proxy                $taxCalculation
+     */
+    public function __construct(
+        ScopeConfigInterface $scopeConfig,
+        Cart $cart,
+        Session $customerSession,
+        GroupRepository $groupRepository,
+        Proxy $taxCalculation
```
Comment (sprankhub) [inline Plugin/Tax/Config/ShippingTaxPlugin.php:60]: We should not depend on a proxy.

**Review by sprankhub — CHANGES_REQUESTED** [review]

**Comment by sprankhub** [comment]
@BorisovskiP, please also check the failing build (probably code style issues).

File: Model/Config.php [hunk 8]
```diff
@@ -188,7 +191,17 @@ public function getCmsBlocks()
...
     }
+
+    /**
+     * Method for retrieving dynamic config shipping path
+     *
+     * @return string
+     */
+    public function getDynamicShippingConfigPath(): string
+    {
+        return self::CONFIG_PATH_DYNAMIC_SHIPPING_TAX_CLASS;
```
Comment (sprankhub) [inline Model/Config.php:205]: I do not think this adds much value. I think we should add a new method in `\FireGento\MageSetup\Model\System\Config`, which returns the actual configuration value of that option.

File: Plugin/Tax/Config/ShippingTaxPlugin.php [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+
+    /**
+     * Method for getting highest product tax class id
+     *
+     * @param array                      $quoteItems
+     * @param null|string|bool|int|Store $store
+     *
+     * @return bool|mixed|null
+     */
+
```
Comment (sprankhub) [inline Plugin/Tax/Config/ShippingTaxPlugin.php:127]: Please remove this line.

File: composer.json [hunk 10]
```diff
@@ -2,10 +2,16 @@
   "name": "firegento/magesetup2",
   "description": "MageSetup provides the necessary configuration (system config, tax, agreements, etc. for a national market.",
   "require": {
-    "php": "~7.2.0||~7.3.0",
-    "magento/module-store": "*",
-    "magento/module-backend": "*",
-    "magento/framework": "*"
+    "php": "~7.2.0|~7.3.0",
```
Comment (sprankhub) [inline composer.json:5]: Please fix the formatting.

File: i18n/de_DE.csv (hunk 1 above)
Comment (sprankhub) [reply i18n/de_DE.csv:58]: It seems that you re-generated the English file. Great! Now IMHO, we need to update the German file accordingly (add the strings to the position similar to the EN file, not at the end of the file).

File: Model/System/Config/Source/Tax/Dynamic.php [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+namespace FireGento\MageSetup\Model\System\Config\Source\Tax;
+
+use Magento\Framework\Data\OptionSourceInterface;
+
+/**
+ * Source model for Dynamic taxes.
+ */
+class Dynamic implements OptionSourceInterface
+{
+    public const DYNAMIC_TYPE_SHIPPING_TAX_DEFAULT = 0;
```
Comment (sprankhub) [inline Model/System/Config/Source/Tax/Dynamic.php:16]: Maybe `DYNAMIC_SHIPPING_TAX_NO`?

File: Model/System/Config/Source/Tax/Dynamic.php [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+    public const DYNAMIC_TYPE_HIGHEST_PRODUCT_TAX = 1;
```
Comment (sprankhub) [inline Model/System/Config/Source/Tax/Dynamic.php:17]: Maybe `DYNAMIC_SHIPPING_TAX_HIGHEST_PRODUCT_TAX`?

**Review by sprankhub — CHANGES_REQUESTED** [review]

File: Plugin/Tax/Config/ShippingTaxPlugin.php [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+     * Method for getting tax prcentage
+     *
+     * @param int                        $productTaxClassId
+     * @param null|string|bool|int|Store $store
+     *
+     * @return int
+     * @throws \Magento\Framework\Exception\LocalizedException
+     * @throws \Magento\Framework\Exception\NoSuchEntityException
+     */
+    private function getTaxPercent(int $productTaxClassId, $store): int
```
Comment (barryvdh) [inline Plugin/Tax/Config/ShippingTaxPlugin.php:176]: The returnType of the rate is float, not int

File: Plugin/Tax/Config/ShippingTaxPlugin.php [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+    {
+        $groupId            = $this->customerSession->getCustomerGroupId();
+        $group              = $this->groupRepository->getById($groupId);
+        $customerTaxClassId = $group->getTaxClassId();
+
+        $request = $this->taxCalculation->getRateRequest(null, null, $customerTaxClassId, $store);
```
Comment (barryvdh) [inline Plugin/Tax/Config/ShippingTaxPlugin.php:182]: This is not actually set in the constructor.

**Comment by barryvdh** [comment]
I'm using a module based on this, but I encounter a loop for the payment methods. I think the collections of the tax percentage might trigger the plugin again, so I added a static variable that prevents loops (doesn't run on nested calls). Not sure if thats always required.

Also it seems the lower method is never called to get the tax percentage. At least it's not functioning right now, not sure if the fallback is needed at all?

**Comment by gino2014** [comment]
Hi Magesetup2 Team and all who are looking for a solution,

as a suggestion you could use this extension

https://github.com/JaJuMa-GmbH/magento-2-dynamic-shipping-tax

File: Plugin/Tax/Config/ShippingTaxPlugin.php (hunk 7 above)
Comment (frostblogNet) [reply Plugin/Tax/Config/ShippingTaxPlugin.php]: The Plugin does not have a around method. I think the plugin will not work in general.
